metrics/core/historical_types.go

```diff
@@ -0,0 +1,153 @@
+
+// AggregationValue is a description of aggregated MetricValues over time
+type AggregationValue struct {
+	Count *uint64
+
+	Aggregations map[string]MetricValue
+}
+
+// TimestampedAggregationValue is an aggregation value with an associated timestamp
+// and bucket size
+type TimestampedAggregationValue struct {
+	Timestamp  time.Time
+	BucketSize time.Duration
+	AggregationValue
+}
+
+// HistoricalKey is an identifier pointing to a particular object.
+// Is is composed of an object type (pod, namespace, container, etc) as well
+// as a series of fields which identify that object.
+type HistoricalKey struct {
+	// ObjectType specifies which type of object this is for (pod, namespace, etc)
+	// It should be one of the MetricSetType* labels.
+	ObjectType string
+
+	// NodeName is used for node and system-container metrics
+	NodeName string
+	// NamespaceName is used for namespace, pod, and pod-container metrics
+	NamespaceName string
+	// PodName is used for pod and pod-container metrics
+	PodName string
+	// ContainerName is used for system-container and pod-container metrics
+	ContainerName string
+	// PodId may be used in place of the combination of PodName and NamespaceName for pod and pod-container metrics
+	PodId string
+}
+
+func (key *HistoricalKey) String() string {
+	prefix := fmt.Sprintf("(%s)", key.ObjectType)
+
+	var path string = "[unknown type]"
+	switch key.ObjectType {
+	case MetricSetTypeSystemContainer:
+		path = fmt.Sprintf("node:%s/container:%s", key.NodeName, key.ContainerName)
+	case MetricSetTypePodContainer:
+		if key.PodId != "" {
+			path = fmt.Sprintf("poduid:%s/container:%s", key.PodId, key.ContainerName)
+		} else {
+			path = fmt.Sprintf("ns:%s/pod:%s/container:%s", key.NamespaceName, key.PodName, key.ContainerName)
+		}
+	case MetricSetTypePod:
+		if key.PodId != "" {
+			path = fmt.Sprintf("poduid:%s", key.PodId)
+		} else {
+			path = fmt.Sprintf("ns:%s/pod:%s", key.NamespaceName, key.PodName)
+		}
+	case MetricSetTypeNamespace:
+		path = fmt.Sprintf("ns:%s", key.NamespaceName)
+	case MetricSetTypeNode:
+		path = fmt.Sprintf("node:%s", key.NodeName)
+	case MetricSetTypeCluster:
+		path = "[cluster]"
+	}
+
+	return prefix + path
+}
+
+// HistoricalSource allows for retrieval of historical metrics and aggregations from sinks
+type HistoricalSource interface {
+	// GetMetric retrieves the given metric for one or more objects (specified by metricKeys) of
+	// the same type, within the given time interval
+	GetMetric(metricName string, metricKeys []HistoricalKey, start, end time.Time) (map[HistoricalKey][]TimestampedMetricValue, error)
+
+	// GetAggregation fetches the given aggregations for one or more objects (specified by metricKeys) of
+	// the same type, within the given time interval, calculated over a series of buckets
+	GetAggregation(metricName string, aggregations []string, metricKeys []HistoricalKey, start, end time.Time, bucketSize time.Duration) (map[HistoricalKey][]TimestampedAggregationValue, error)
```

Comment:
Just a type alias around string?  Sure.  I often think that Go could use a proper enum type, though...